### PR TITLE
HPCC-18780 Generate projected as well as expected disk meta functions

### PR DESCRIPTION
**Pull request by richardkchapman**
Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
Compiled

### CODE REVIEW

**Comment by richardkchapman** [comment]
@ghalliday Please review

**Comment by richardkchapman** [comment]
@jakesmith Please note/review the lurking thor bug that this change revealed. You may prefer to fix it differently (not sure why you don't overwrite basehelper in the container before construction for example).

There is similar code on the master side, though I don't think it actually makes any calls through the helper before it is updated, it's potentially an accident waiting to happen still.

I don't know what if any ill effects this might be causing in pre 7.0 versions and whether worth backporting the fix to them.

File: thorlcr/activities/thdiskbaseslave.cpp [hunk 1]
```diff
@@ -202,8 +202,10 @@ const char * CDiskPartHandlerBase::queryLogicalFilename(const void * row)
 
 //////////////////////////////////////////////
 
-CDiskReadSlaveActivityBase::CDiskReadSlaveActivityBase(CGraphElementBase *_container) : CSlaveActivity(_container)
+CDiskReadSlaveActivityBase::CDiskReadSlaveActivityBase(CGraphElementBase *_container, IHThorArg *_helper) : CSlaveActivity(_container)
 {
+    if (_helper)
+        baseHelper.set(_helper);
     helper = (IHThorDiskReadBaseArg *)queryHelper();
     reInit = 0 != (helper->getFlags() & (TDXvarfilename|TDXdynamicfilename));
```
Comment (richardkchapman) [inline thorlcr/activities/thdiskbaseslave.cpp]: Prior to my change, this might act on the original WorkUnitReadArg

**Comment by jakesmith** [comment]
@richardkchapman 
>not sure why you don't overwrite basehelper in the container before construction for example

Nor am I. I think the container itself should be converted, i.e. the detection should happen much earlier, so there's no potential confusion downstream in either master or slave.

>There is similar code on the master side, though I don't think it actually makes any calls through the helper before it is updated, it's potentially an accident waiting to happen still.

I've checked and it's safe, but as you a potential problem to trip over in the future.

However, I think can go with this change and I'll open a new JIRA to look at refactoring the conversion so it happens earlier.

@ghalliday - can you review the rest ?

**Pull request merged**